### DISCUSSION TITLE
How to unwarp the foveated frames outside alvr?

### DISCUSSION THREAD

**Discussion by Jia-WeiFang** [alvr-org/ALVR#1311] [Q&A]
Hi, 
**Since I want to calculate the objective metrics between the foveated-wrapped images and unwarped images, I save both the original eye texture and the warped texture on the server side.**
(The reason why I didn't just save frames from the client is that the HMD hardware can not afford to save frames in runtime, the workload is too heavy; the app would shut down. In addition, the HMD will do some cropping before displaying the frames, so I can not directly use the screen recording either.)

Thus, **I wonder how can I do the unwarping function offline?**
I have created an OpenGL project and followed the render pipeline in the client source code, and use "CalculateFoveationVars" function to calculate the foveated parameters, the fragment shader in alvr/client_core/cpp/ffr.cpp, and the vertex shader "QUAD_2D_VERTEX_SHADER", but I can not get the correct unwarped results.

**Can I have the original test OpenGL project of doing the warping and unwarping work, or can you tell me how to do this offline?**

Thanks a lot!
Really appreciate your help!

Jia-Wei

**Comment by zmerp** [comment]
The current FFR algorithm was written by @sctanf, maybe she can help. if you want to record the stream output I highly recommend to just not using FFR.

**Reply by Jia-WeiFang** [reply]
Hi, thanks for your reply!
The reason why I need to do this is that my research problem is mainly about "foveated" streaming, so I need to capture the frames with ffr on.

@sctanf How can I contact you?
Thanks a lot for your help!!!

Jia-Wei

**Reply by zmerp** [reply]
@Jia-WeiFang At the bottom of this page https://github.com/alvr-org/ALVR/wiki/How-ALVR-works you can find two shadertoy examples written by me before implementing the Warp and Slices FFR methods. The method that ALVR currently uses is an adaptation of Oculus' AADT (Axis Aligned Distorted Transfer). You can look in the commit history to find the code relative to the first two methods.

**Reply by Jia-WeiFang** [reply]
@zarik5 
Hi, thanks again for your reply!
I have checked the page you said and saw your demo on shadertoy.
However, I need the FFR methods used currently in ALVR, so I may have to contact the author @sctanf.
Do you know how to contact her?

Thanks a lot for your help!!!!

Jia-Wei

**Reply by sctanf** [reply]
@Jia-WeiFang current ffr method on the code is the only working example of it
I can link a graph that shows the warping currently used but it and the shadertoy may be missing alignment logic; The actual shader in ALVR aligns to the nearest pixel and nearest 32px resolution iirc
https://www.desmos.com/calculator/cmvjr7ljje
https://www.shadertoy.com/view/fd3SW2

**Reply by Jia-WeiFang** [reply]
@sctanf Hi!
Thanks for your kind reply and sharing the graph and shadertoy to me.

So if I want to do the unwarping offline, is it right to directly use **the vertex shader QUAD_2D_VERTEX_SHADER** in render_pipeline.h and **the fragment shader FFR_COMMON_SHADER_FORMAT + DECOMPRESS_AXIS_ALIGNED_FRAGMENT_SHADER** in ffr.cpp?

Or should I do anything else to the shader?

So far I think my opengl project can run.
However, I think the uv coordinate in _color = texture(tex0, **EyeToTextureUV(uncompressedUV * EYE_SIZE_RATIO, isRightEye)**);_ is wrong because if I change this line to _color = texture(tex0, uv));_, the program can render the upside-down image correctly(this result also show that the vertex shader can work).

**Reply by sctanf** [reply]
@Jia-WeiFang color=texture(tex0, uv); would skip the decompressing entirely

**Reply by Jia-WeiFang** [reply]
Dear @sctanf,
May I ask what are the meanings of your non-linear and linear functions in **desmos** ?

Thanks a lot for your help!
Jia-Wei

**Reply by sctanf** [reply]
@Jia-WeiFang the non-linear sections have increasing compression as it reaches the edge, the linear section is uncompressed

**Reply by Jia-WeiFang** [reply]
Dear @sctanf,
Thanks for your reply!

Does your non-linear function follow some math theory?
Or it is just your heuristic method?

Can you briefly explain your design concept~?

Thank you a lot!!!

Jia-Wei

**Reply by sctanf** [reply]
@Jia-WeiFang the functions are created to have an adjustable average slope and avoid discontinuity but there is not much other consideration in its design, other than some pixel alignment in the final shader
